### PR TITLE
Calendar Module QOL Features and Adjustments

### PR DESCRIPTION
**Pull request by CarJem**
This commit adds several QOL features and adjustments to the calendar module including: 
- **New Options**
    - ``coloredText``: ``(default: false)`` Determines if you want your entry text to be colored based on the calendar's color
    - ``coloredBorder``: ``(default: false)`` Determines if you want entry borders to be colored based on the calendar's color 
    - ``coloredSymbol``: ``(default: false)`` Determines if you want entry symbols to be colored based on the calendar's color
    - ``coloredBackground``: ``(default: false)`` Determines if you want entry backgrounds to be colored based on the calendar's color
        > These new colored options allows for more out-of-box styling options for the calendar module. With this the ``coloredSymbolOnly`` option has been removed due to redundancy
    - ``limitDaysNeverSkip``: ``(default: false)`` show every event for every day regardless of if the day only has a single full day event  
    - ``flipDateHeaderTitle``: ``(default: false)`` determines if the title for the date header in the ``dateheaders`` time format should align to the left ``[eg: false]`` or right ``[eg: true]``
- **Layout Changes** 
    - ``dateheader`` is now a class avaliable for date headers in the ``dateheaders`` time format.
    - Event entries have been better *container-ized* for better styling (using the ``event-container`` class)
    - ``repeatingCountTitle`` now has a seperator between the ``yearDiff`` and ``repeatingCountTitle``
    - ``endDate`` in ``dateheaders`` now capitalizes it's first letter

### CODE REVIEW

File: modules/default/calendar/calendar.js [hunk 1]
```diff
@@ -40,7 +40,6 @@ Module.register("calendar", {
 		hideTime: false,
 		showTimeToday: false,
 		colored: false,
-		coloredSymbolOnly: false,
```
Comment (rejas) [inline modules/default/calendar/calendar.js:43]: I dont like removing config options without a) telling the users beforehand and b) try to provide an autoamtic replacement in code, see https://github.com/MichMich/MagicMirror/blob/4b478a5a5e891ea493171b7867ee473a15cb86cc/modules/default/weather/weather.js#L80 for an example

File: modules/default/calendar/calendar.js [hunk 2]
```diff
@@ -744,6 +765,16 @@ Module.register("calendar", {
 		return this.getCalendarProperty(url, "color", "#fff");
 	},
 
+	/**
+	 * Retrieves the bgColor for a specific calendar url.
+	 *
+	 * @param {string} url The calendar url
+	 * @returns {string} The color
+	 */
+	bgColorForUrl: function (url) {
```
Comment (rejas) [inline modules/default/calendar/calendar.js:774]: this is only used in one place, why create a method for this then?

**Review by rejas — CHANGES_REQUESTED** [review]

File: modules/default/calendar/calendar.js (hunk 2 above)
Comment (CarJem) [reply modules/default/calendar/calendar.js:774]: I'll make adjustments and respectively update my code

File: modules/default/calendar/calendar.css [hunk 3]
```diff
@@ -1,7 +1,4 @@
 .calendar .symbol {
-  display: flex;
```
[the remark this answers is not in the document]
Comment (CarJem) [reply modules/default/calendar/calendar.css:2]: Because the flex causes this to screw up and generate extra empty space
image

[the remark this answers is not in the document]
Comment (CarJem) [reply modules/default/calendar/calendar.css:2]: I can put it back... but it wouldn't look pleasing on the eyes

**Comment by CarJem** [comment]
I have addressed all of the issues, minus the style issue, which I believe is necessary to ensure that most layouts don't break under these changes.

File: modules/default/calendar/calendar.css (hunk 3 above)
[the remark this answers is not in the document]
Comment (rejas) [reply modules/default/calendar/calendar.css:2]: And like I said it would break existing layouts, like mine:

grafik

It might be necessary to cleanup the layout of the calendar module, but not in this PR :-)

File: modules/default/calendar/calendar.js [hunk 4]
```diff
@@ -40,7 +40,6 @@ Module.register("calendar", {
 		hideTime: false,
 		showTimeToday: false,
 		colored: false,
```
Comment (rejas) [inline modules/default/calendar/calendar.js:42]: the colored option is ignored now with your PR. You should add some fallback logic for that in the depreaction warning block, so that existing configs work like before.

File: modules/default/calendar/calendar.css (hunk 3 above)
[the remark this answers is not in the document]
Comment (CarJem) [reply modules/default/calendar/calendar.css:2]: Very well then, I will put the changes back and make a temporary work around for myself... I don't want to hold of this PR any longer than is needed

File: modules/default/calendar/calendar.js [hunk 5]
```diff
@@ -133,14 +142,6 @@ Module.register("calendar", {
 			// fetcher till cycle
 			this.addCalendar(calendar.url, calendar.auth, calendarConfig);
 		});
-
```
Comment (rejas) [inline modules/default/calendar/calendar.js:136]: Why did you remove that? Oversight I guess?

**Review by rejas — CHANGES_REQUESTED** [review]

File: modules/default/calendar/calendar.css (hunk 3 above)
[the remark this answers is not in the document]
Comment (rejas) [reply modules/default/calendar/calendar.css:2]: Thanks. I do like your PR and the screenshot looks nice, but we got to take care of exisiting users. Thanks for you understanding

File: modules/default/calendar/calendar.js (hunk 5 above)
Comment (CarJem) [reply modules/default/calendar/calendar.js]: I didn't find it in the latest commit so I removed it, I'll put it back

File: modules/default/calendar/calendar.js (hunk 4 above)
Comment (CarJem) [reply modules/default/calendar/calendar.js:42]: Done on latest commit.

**Comment by CarJem** [comment]
P.S. If you would like to speed this up I can be reached via Discord.

File: modules/default/calendar/calendar.js [hunk 6]
```diff
@@ -678,6 +710,7 @@ Module.register("calendar", {
 					// Get the default prefix for this class name and add to the custom symbol provided
 					const className = this.getCalendarProperty(event.url, "symbolClassName", this.config.defaultSymbolClassName);
 					symbols[0] = className + ev.symbol;
+					symbols[0] = ev.symbol;
```
Comment (rejas) [inline modules/default/calendar/calendar.js:713]: this seems buggy, since you overwrite the line before...

**Review by rejas — CHANGES_REQUESTED** [review]
just one small thing...

File: modules/default/calendar/calendar.js (hunk 6 above)
Comment (CarJem) [reply modules/default/calendar/calendar.js]: Yeah, I just pushed a fix

**Review by rejas — APPROVED** [review]

**Review by rejas — CHANGES_REQUESTED** [review]
like we say in german: "zu früh gefreut". found a bug ....

**Comment by CarJem** [comment]
Okay, how do I resolve these checks?

**Comment by CarJem** [comment]
I guess I'll start by running prettier again

**Comment by rejas** [comment]
> Okay, how do I resolve these checks?

Do
`npm run lint:prettier`
and commit the changes to fix the tests.

**Review by CarJem — COMMENTED** [review]
Done

**Comment by rejas** [comment]
Grmpf, some other changes (not the css ones) were crashing my layout. Will look into it...

**Comment by CarJem** [comment]
I was gonna put something in the changelog... but I don't know how to summarize this nicely

**Comment by CarJem** [comment]
> Grmpf, some other changes (not the css ones) were crashing my layout. Will look into it...

Okay, currently working on not alot of time... Once the hour is up, I must return to my day job... hopefully we can still make traction

**Comment by rejas** [comment]
Dinner time here... Seems the newly added event-tables are the culprit...

**Comment by CarJem** [comment]
> Dinner time here... Seems the newly added event-tables are the culprit...

I'd like to refer to them as eventContainers please, anyways, I apologize, I had to merge this from an old build into the present and it hasn't been easy to do as you can tell...

**Comment by CarJem** [comment]
> Dinner time here... Seems the newly added event-tables are the culprit...

If you can provide me some logs I can diagnose

**Comment by rejas** [comment]
You should get rid of the eventContainer and add your new styles to the eventWrapper. Then it looks like before. Would that break your personal layout?

**Comment by CarJem** [comment]
> You should get rid of the eventContainer and add your new styles to the eventWrapper. Then it looks like before. Would that break your personal layout?

Yes, and along with the new control for styling too

**Comment by rejas** [comment]
> I was gonna put something in the changelog... but I don't know how to summarize this nicely

"Added new calendar options for colored entries: ....." sounds nice :-)

**Comment by rejas** [comment]
> Yes, and along with the new control for styling too

But the existing layout depend on the table strcuture so that icons / title /time are aligned correctly. Your new eventContainer div breaks up the table and therefore breaks the exisiting layouts. Now off to dinner, sorry

**Comment by CarJem** [comment]
> > Yes, and along with the new control for styling too
> 
> But the existing layout depend on the table strcuture so that icons / title /time are aligned correctly. Your new eventContainer div breaks up the table and therefore breaks the exisiting layouts. Now off to dinner, sorry

I'll see what can be changed then.... ugh....

**Comment by CarJem** [comment]
Surprisingly, this should make both kinds of layouts work with these changes

**Comment by CarJem** [comment]
> Surprisingly, this should make both legacy and new layouts work with these changes. The only questionable change that is required for this to work was moving the location into the eventWrapper.

Of which I just reverted and properly implemented

**Comment by rejas** [comment]
Merging it finally, thanks for your patience.

Would be cool if you could prepare a PR for the documentation (https://github.com/MichMich/MagicMirror-Documentation) with the new config options.

Aaaand if you really got time (soo optional), maybe write tests for the new options? ;-)
